src/picongpu/include/particles/ionization/ionization.hpp

```diff
@@ -0,0 +1,386 @@
+    const DataSpace<simDim> blockCell = block * SuperCellSize::toRT();
+    
+    typedef typename particles::ionization::WriteElectronIntoFrame WriteElectronIntoFrame;
+
+    __syncthreads();
+
+    __shared__ IONFRAME *ionFrame;
+    __shared__ ELECTRONFRAME *electronFrame;
+    __shared__ bool isValid;
+    __shared__ lcellId_t maxParticlesInFrame;
+
+    __syncthreads(); /*wait that all shared memory is initialized*/
+
+    /* find last frame in super cell 
+     * define maxParticlesInFrame as the maximum frame size */
+    if (linearThreadIdx == 0)
+    {
+        ionFrame = &(ionBox.getLastFrame(block, isValid));
+        maxParticlesInFrame = PMacc::math::CT::volume<SuperCellSize>::type::value;
+    }
+
+    __syncthreads();
+    if (!isValid)
+        return; //end kernel if we have no frames
+
+    /* caching of E and B fields */
+    PMACC_AUTO(cachedB, CachedBox::create < 0, typename BBox::ValueType > (BlockDescription_()));
+    PMACC_AUTO(cachedE, CachedBox::create < 1, typename EBox::ValueType > (BlockDescription_()));
+
+    __syncthreads();
+    
+    nvidia::functors::Assign assign;
+    
+    PMACC_AUTO(fieldBBlock, fieldB.shift(blockCell));
+    ThreadCollective<BlockDescription_> collectiv(linearThreadIdx);
+    collectiv(
+              assign,
+              cachedB,
+              fieldBBlock
+              );
+
+    PMACC_AUTO(fieldEBlock, fieldE.shift(blockCell));
+    collectiv(
+              assign,
+              cachedE,
+              fieldEBlock
+              );
+    
+    /* Declare counter in shared memory that will later tell the current fill level or 
+     * occupation of the newly created target electron frames. */
+    __shared__ int newFrameFillLvl;
+    
+    __syncthreads(); /*wait that all shared memory is initialized*/
+    
+    /* Declare local variable oldFrameFillLvl for each thread */
+    int oldFrameFillLvl;
+    
+    /* Initialize local (register) counter for each thread 
+     * - describes how many new electrons should be created */
+    unsigned int newElectrons = 0;
+    
+    /* Declare local electron ID
+     * - describes at which position in the new frame the new electron is to be created */
+    int electronId;
+    
+    /* Master initializes the frame fill level with 0 */
+    if (linearThreadIdx == 0)
+    {
+        newFrameFillLvl = 0;        
+        electronFrame = NULL;
+    }
+    __syncthreads();
+    
+    /* move over particle frames and call frameIonizer 
+     * frames are worked on in backwards order to avoid asking if there is another frame
+     * --> performance 
+     * Because all frames are completely filled except the last and apart from that last frame 
+     * one wants to make sure that all threads are working and every frame is worked on. */
+    while (isValid)
+    {
+        /* casting uint8_t multiMask to boolean */
+        bool isParticle = (*ionFrame)[linearThreadIdx][multiMask_];
+        __syncthreads();;
```

Comment:
Yeah, now you're making fun of this, you two. But what if I this layout is a key feature in my Feng Shui arrangement of the code? Removing this would result in serious unbalancing of the Qi...